File: bindings/rust/src/triplestore.rs
// SPDX-License-Identifier: MPL-2.0
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk>
//

//! Triple Store types for the proven-servers ABI.
//!
//! Formally verified RDF triple store types.
//! Mirrors the Idris2 module `TriplestoreABI.Types`.
//!
//! - `Statement` -- RDF statement types.
//! - `IndexOrder` -- Triple index orderings.
//! - `StorageBackend` -- Triple store storage backends.
//! - `ImportFormat` -- RDF import formats.
//! - `TransactionIsolation` -- Triple store transaction isolation.
//! - `StoreState` -- Triple store states.
//!
//! All discriminant values match the Idris2 ABI tag definitions exactly.

use std::fmt;

// ===========================================================================
// Statement (tags 0-1)
// ===========================================================================

/// RDF statement types.
///
/// Matches `Statement` in `TriplestoreABI.Types`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
#[repr(u8)]
pub enum Statement {
    /// Triple (tag 0).
    Triple = 0,
    /// Quad (tag 1).
    Quad = 1,
}

impl Statement {
    /// Decode from an ABI tag value.
    pub fn from_tag(tag: u8) -> Option<Self> {
        match tag {
            0 => Some(Self::Triple),
            1 => Some(Self::Quad),
            _ => None,
        }
    }

    /// Encode to the ABI tag value.
    pub fn to_tag(self) -> u8 {
        self as u8
    }

    /// All variants of this type.
    pub const ALL: [Statement; 2] = [
        Self::Triple, Self::Quad,
    ];
}

impl fmt::Display for Statement {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Debug::fmt(self, f)
    }
}

// ===========================================================================
// IndexOrder (tags 0-5)
// ===========================================================================

/// Triple index orderings.
///
/// Matches `IndexOrder` in `TriplestoreABI.Types`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
#[repr(u8)]
pub enum IndexOrder {
    /// SPO (tag 0).
    Spo = 0,
    /// POS (tag 1).
    Pos = 1,
    /// OSP (tag 2).
    Osp = 2,
    /// GSPO (tag 3).
    Gspo = 3,
    /// GPOS (tag 4).
    Gpos = 4,
    /// GOSP (tag 5).
    Gosp = 5,
}

impl IndexOrder {
    /// Decode from an ABI tag value.
    pub fn from_tag(tag: u8) -> Option<Self> {
        match tag {
            0 => Some(Self::Spo),
            1 => Some(Self::Pos),
            2 => Some(Self::Osp),
            3 => Some(Self::Gspo),
            4 => Some(Self::Gpos),
            5 => Some(Self::Gosp),
            _ => None,
        }
    }

    /// Encode to the ABI tag value.
    pub fn to_tag(self) -> u8 {
        self as u8
    }

    /// All variants of this type.
    pub const ALL: [IndexOrder; 6] = [
        Self::Spo, Self::Pos, Self::Osp, Self::Gspo, Self::Gpos, Self::Gosp,
    ];
}

impl fmt::Display for IndexOrder {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Debug::fmt(self, f)
    }
}

// ===========================================================================
// StorageBackend (tags 0-3)
// ===========================================================================

/// Triple store storage backends.
///
/// Matches `StorageBackend` in `TriplestoreABI.Types`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
#[repr(u8)]
pub enum StorageBackend {
    /// InMemory (tag 0).
    InMemory = 0,
    /// BTree (tag 1).
    BTree = 1,
    /// LSM (tag 2).
    Lsm = 2,
    /// Persistent (tag 3).
    Persistent = 3,
}

impl StorageBackend {
    /// Decode from an ABI tag value.
    pub fn from_tag(tag: u8) -> Option<Self> {
        match tag {
            0 => Some(Self::InMemory),
            1 => Some(Self::BTree),
            2 => Some(Self::Lsm),
            3 => Some(Self::Persistent),
            _ => None,
        }
    }

    /// Encode to the ABI tag value.
    pub fn to_tag(self) -> u8 {
        self as u8
    }

    /// All variants of this type.
    pub const ALL: [StorageBackend; 4] = [
        Self::InMemory, Self::BTree, Self::Lsm, Self::Persistent,
    ];
}

impl fmt::Display for StorageBackend {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Debug::fmt(self, f)
    }
}

// ===========================================================================
// ImportFormat (tags 0-5)
// ===========================================================================

/// RDF import formats.
///
/// Matches `ImportFormat` in `TriplestoreABI.Types`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
#[repr(u8)]
pub enum ImportFormat {
    /// NTriples (tag 0).
    NTriples = 0,
    /// Turtle (tag 1).
    Turtle = 1,
    /// RDF/XML (tag 2).
    RdfXml = 2,
    /// JSON-LD (tag 3).
    JsonLd = 3,
    /// NQuads (tag 4).
    NQuads = 4,
    /// Trig (tag 5).
    Trig = 5,
}

impl ImportFormat {
    /// Decode from an ABI tag value.
    pub fn from_tag(tag: u8) -> Option<Self> {
        match tag {
            0 => Some(Self::NTriples),
            1 => Some(Self::Turtle),
            2 => Some(Self::RdfXml),
            3 => Some(Self::JsonLd),
            4 => Some(Self::NQuads),
            5 => Some(Self::Trig),
            _ => None,
        }
    }

    /// Encode to the ABI tag value.
    pub fn to_tag(self) -> u8 {
        self as u8
    }

    /// All variants of this type.
    pub const ALL: [ImportFormat; 6] = [
        Self::NTriples, Self::Turtle, Self::RdfXml, Self::JsonLd, Self::NQuads, Self::Trig,
    ];
}

impl fmt::Display for ImportFormat {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Debug::fmt(self, f)
    }
}

// ===========================================================================
// TransactionIsolation (tags 0-2)
// ===========================================================================

/// Triple store transaction isolation.
///
/// Matches `TransactionIsolation` in `TriplestoreABI.Types`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
#[repr(u8)]
pub enum TransactionIsolation {
    /// ReadCommitted (tag 0).
    ReadCommitted = 0,
    /// Serializable (tag 1).
    Serializable = 1,
    /// Snapshot (tag 2).
    Snapshot = 2,
}

impl TransactionIsolation {
    /// Decode from an ABI tag value.
    pub fn from_tag(tag: u8) -> Option<Self> {
        match tag {
            0 => Some(Self::ReadCommitted),
            1 => Some(Self::Serializable),
            2 => Some(Self::Snapshot),
            _ => None,
        }
    }

    /// Encode to the ABI tag value.
    pub fn to_tag(self) -> u8 {
        self as u8
    }

    /// All variants of this type.
    pub const ALL: [TransactionIsolation; 3] = [
        Self::ReadCommitted, Self::Serializable, Self::Snapshot,
    ];
}

impl fmt::Display for TransactionIsolation {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Debug::fmt(self, f)
    }
}

// ===========================================================================
// StoreState (tags 0-4)
// ===========================================================================

/// Triple store states.
///
/// Matches `StoreState` in `TriplestoreABI.Types`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
#[repr(u8)]
pub enum StoreState {
    /// Idle (tag 0).
    Idle = 0,
    /// Ready (tag 1).
    Ready = 1,
    /// In transaction (tag 2).
    InTransaction = 2,
    /// Importing (tag 3).
    Importing = 3,
    /// Closing (tag 4).
    Closing = 4,
}

impl StoreState {
    /// Decode from an ABI tag value.
    pub fn from_tag(tag: u8) -> Option<Self> {
        match tag {
            0 => Some(Self::Idle),
            1 => Some(Self::Ready),
            2 => Some(Self::InTransaction),
            3 => Some(Self::Importing),
            4 => Some(Self::Closing),
            _ => None,
        }
    }

    /// Encode to the ABI tag value.
    pub fn to_tag(self) -> u8 {
        self as u8
    }

    /// All variants of this type.
    pub const ALL: [StoreState; 5] = [
        Self::Idle, Self::Ready, Self::InTransaction, Self::Importing, Self::Closing,
    ];
}

impl fmt::Display for StoreState {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Debug::fmt(self, f)
    }
}

// ===========================================================================
// Tests
// ===========================================================================

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn statement_roundtrip() {
        for v in Statement::ALL {
            let tag = v.to_tag();
            let decoded = Statement::from_tag(tag).expect("valid tag");
            assert_eq!(decoded, v);
        }
        assert!(Statement::from_tag(2).is_none());
    }

    #[test]
    fn index_order_roundtrip() {
        for v in IndexOrder::ALL {
            let tag = v.to_tag();
            let decoded = IndexOrder::from_tag(tag).expect("valid tag");
            assert_eq!(decoded, v);
        }
        assert!(IndexOrder::from_tag(6).is_none());
    }

    #[test]
    fn storage_backend_roundtrip() {
        for v in StorageBackend::ALL {
            let tag = v.to_tag();
            let decoded = StorageBackend::from_tag(tag).expect("valid tag");
            assert_eq!(decoded, v);
        }
        assert!(StorageBackend::from_tag(4).is_none());
    }

    #[test]
    fn import_format_roundtrip() {
        for v in ImportFormat::ALL {
            let tag = v.to_tag();
            let decoded = ImportFormat::from_tag(tag).expect("valid tag");
            assert_eq!(decoded, v);
        }
        assert!(ImportFormat::from_tag(6).is_none());
    }

    #[test]
    fn transaction_isolation_roundtrip() {
        for v in TransactionIsolation::ALL {
            let tag = v.to_tag();
            let decoded = TransactionIsolation::from_tag(tag).expect("valid tag");
            assert_eq!(decoded, v);
        }
        assert!(TransactionIsolation::from_tag(3).is_none());
    }

    #[test]
    fn store_state_roundtrip() {
        for v in StoreState::ALL {
            let tag = v.to_tag();
            let decoded = StoreState::from_tag(tag).expect("valid tag");
            assert_eq!(decoded, v);
        }
        assert!(StoreState::from_tag(5).is_none());
    }

}

File: protocols/proven-triplestore/ffi/zig/src/triplestore.zig
// SPDX-License-Identifier: MPL-2.0
// Copyright (c) Jonathan D.A. Jewell <j.d.a.jewell@open.ac.uk>
//
// triplestore.zig -- Zig FFI implementation of proven-triplestore.
//
// Implements the RDF triple store state machine with:
//   - 64-slot mutex-protected store session pool
//   - Triple/quad insertion and deletion
//   - Pattern-based triple lookup (exact match)
//   - Transaction begin/commit/rollback
//   - Bulk import lifecycle
//   - Store state machine (Idle -> Ready -> Transaction/Importing -> Closing)
//   - Thread-safe via per-pool mutex
//
// All exported functions use C calling convention (callconv(.c)) and
// communicate state via u8 tags matching TriplestoreABI.Types exactly.

const std = @import("std");

// =========================================================================
// Enums (matching TriplestoreABI.Types tag assignments)
// =========================================================================

/// RDF statement types (ABI tags 0-1).
pub const StatementType = enum(u8) {
    triple = 0,
    quad = 1,
};

/// Index orderings (ABI tags 0-5).
pub const IndexOrder = enum(u8) {
    spo = 0,
    pos = 1,
    osp = 2,
    gspo = 3,
    gpos = 4,
    gosp = 5,
};

/// Storage backends (ABI tags 0-3).
pub const StorageBackend = enum(u8) {
    in_memory = 0,
    btree = 1,
    lsm = 2,
    persistent = 3,
};

/// Import formats (ABI tags 0-5).
pub const ImportFormat = enum(u8) {
    n_triples = 0,
    turtle = 1,
    rdf_xml = 2,
    json_ld = 3,
    n_quads = 4,
    trig = 5,
};

/// Transaction isolation levels (ABI tags 0-2).
pub const TransactionIsolation = enum(u8) {
    read_committed = 0,
    serializable = 1,
    snapshot = 2,
};

/// Store lifecycle states (ABI tags 0-4).
pub const StoreState = enum(u8) {
    idle = 0,
    ready = 1,
    transaction = 2,
    importing = 3,
    closing = 4,
};

// =========================================================================
// Internal data structures
// =========================================================================

/// Maximum concurrent sessions.
const MAX_SESSIONS: usize = 8;

/// Maximum triples per store session.
const MAX_TRIPLES: usize = 256;

/// Maximum URI/literal length.
const MAX_TERM_LEN: usize = 512;

/// An RDF triple (subject, predicate, object).
const Triple = struct {
    subject: [MAX_TERM_LEN]u8,
    subject_len: u32,
    predicate: [MAX_TERM_LEN]u8,
    predicate_len: u32,
    object: [MAX_TERM_LEN]u8,
    object_len: u32,
    /// Optional graph for quads (len=0 means default graph / triple).
    graph: [MAX_TERM_LEN]u8,
    graph_len: u32,
    /// Whether this slot is in use.
    active: bool,
};

/// Default (empty) triple.
const empty_triple: Triple = .{
    .subject = [_]u8{0} ** MAX_TERM_LEN,
    .subject_len = 0,
    .predicate = [_]u8{0} ** MAX_TERM_LEN,
    .predicate_len = 0,
    .object = [_]u8{0} ** MAX_TERM_LEN,
    .object_len = 0,
    .graph = [_]u8{0} ** MAX_TERM_LEN,
    .graph_len = 0,
    .active = false,
};

/// A triple store session.
const Session = struct {
    /// Current store lifecycle state.
    state: StoreState,
    /// Storage backend.
    backend: StorageBackend,
    /// Transaction isolation level.
    isolation: TransactionIsolation,
    /// Current import format (valid only during import).
    import_format: ImportFormat,
    /// Triple storage.
    triples: [MAX_TRIPLES]Triple,
    /// Number of active triples.
    triple_count: u32,
    /// Number of triples at transaction start (for rollback).
    txn_snapshot_count: u32,
    /// Whether this session slot is in use.
    active: bool,
};

/// Default (empty) session.
const empty_session: Session = .{
    .state = .idle,
    .backend = .in_memory,
    .isolation = .read_committed,
    .import_format = .n_triples,
    .triples = [_]Triple{empty_triple} ** MAX_TRIPLES,
    .triple_count = 0,
    .txn_snapshot_count = 0,
    .active = false,
};

// =========================================================================
// Global state
// =========================================================================

var sessions: [MAX_SESSIONS]Session = [_]Session{empty_session} ** MAX_SESSIONS;
var mutex: std.Thread.Mutex = .{};

// =========================================================================
// Internal helpers
// =========================================================================

/// Validate a slot index, returning null if out of range or inactive.
fn validSlot(slot: c_int) ?usize {
    if (slot < 0 or slot >= MAX_SESSIONS) return null;
    const idx: usize = @intCast(slot);
    if (!sessions[idx].active) return null;
    return idx;
}

/// Check if two byte slices are equal.
fn bytesEqual(a: []const u8, b: []const u8) bool {
    if (a.len != b.len) return false;
    return std.mem.eql(u8, a, b);
}

// =========================================================================
// Exported C ABI functions
// =========================================================================

/// Returns the ABI version number.
pub export fn triplestore_abi_version() callconv(.c) u32 {
    return 1;
}

/// Create a new store session. Returns slot index (>=0) or -1 on failure.
pub export fn triplestore_create(backend: u8, isolation: u8) callconv(.c) c_int {
    mutex.lock();
    defer mutex.unlock();

    if (backend > 3) return -1;
    if (isolation > 2) return -1;

    for (&sessions, 0..) |*s, i| {
        if (!s.active) {
            s.* = empty_session;
            s.backend = @enumFromInt(backend);
            s.isolation = @enumFromInt(isolation);
            s.state = .ready;
            s.active = true;
            return @intCast(i);
        }
    }
    return -1;
}

/// Destroy a session, releasing its slot.
pub export fn triplestore_destroy(slot: c_int) callconv(.c) void {
    mutex.lock();
    defer mutex.unlock();
    if (slot < 0 or slot >= MAX_SESSIONS) return;
    sessions[@intCast(slot)] = empty_session;
}

/// Returns the current StoreState tag for a session.
pub export fn triplestore_state(slot: c_int) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();
    const idx = validSlot(slot) orelse return 0;
    return @intFromEnum(sessions[idx].state);
}

/// Add an RDF triple. Returns 0 on success, 1 on rejection.
pub export fn triplestore_add_triple(
    slot: c_int,
    s_ptr: [*]const u8,
    s_len: u32,
    p_ptr: [*]const u8,
    p_len: u32,
    o_ptr: [*]const u8,
    o_len: u32,
) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 1;
    const state = sessions[idx].state;
    if (state != .ready and state != .transaction and state != .importing) return 1;
    if (s_len == 0 or s_len > MAX_TERM_LEN) return 1;
    if (p_len == 0 or p_len > MAX_TERM_LEN) return 1;
    if (o_len == 0 or o_len > MAX_TERM_LEN) return 1;

    // Find a free triple slot
    for (&sessions[idx].triples) |*t| {
        if (!t.active) {
            @memcpy(t.subject[0..s_len], s_ptr[0..s_len]);
            t.subject_len = s_len;
            @memcpy(t.predicate[0..p_len], p_ptr[0..p_len]);
            t.predicate_len = p_len;
            @memcpy(t.object[0..o_len], o_ptr[0..o_len]);
            t.object_len = o_len;
            t.graph_len = 0;
            t.active = true;
            sessions[idx].triple_count += 1;
            return 0;
        }
    }
    return 1; // No free slots
}

/// Add an RDF quad. Returns 0 on success, 1 on rejection.
pub export fn triplestore_add_quad(
    slot: c_int,
    g_ptr: [*]const u8,
    g_len: u32,
    s_ptr: [*]const u8,
    s_len: u32,
    p_ptr: [*]const u8,
    p_len: u32,
    o_ptr: [*]const u8,
    o_len: u32,
) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 1;
    const state = sessions[idx].state;
    if (state != .ready and state != .transaction and state != .importing) return 1;
    if (g_len == 0 or g_len > MAX_TERM_LEN) return 1;
    if (s_len == 0 or s_len > MAX_TERM_LEN) return 1;
    if (p_len == 0 or p_len > MAX_TERM_LEN) return 1;
    if (o_len == 0 or o_len > MAX_TERM_LEN) return 1;

    for (&sessions[idx].triples) |*t| {
        if (!t.active) {
            @memcpy(t.graph[0..g_len], g_ptr[0..g_len]);
            t.graph_len = g_len;
            @memcpy(t.subject[0..s_len], s_ptr[0..s_len]);
            t.subject_len = s_len;
            @memcpy(t.predicate[0..p_len], p_ptr[0..p_len]);
            t.predicate_len = p_len;
            @memcpy(t.object[0..o_len], o_ptr[0..o_len]);
            t.object_len = o_len;
            t.active = true;
            sessions[idx].triple_count += 1;
            return 0;
        }
    }
    return 1;
}

/// Remove a triple by exact match. Returns 0 on success, 1 on not found.
pub export fn triplestore_remove(
    slot: c_int,
    s_ptr: [*]const u8,
    s_len: u32,
    p_ptr: [*]const u8,
    p_len: u32,
    o_ptr: [*]const u8,
    o_len: u32,
) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 1;
    const state = sessions[idx].state;
    if (state != .ready and state != .transaction) return 1;

    const s = s_ptr[0..s_len];
    const p = p_ptr[0..p_len];
    const o = o_ptr[0..o_len];

    for (&sessions[idx].triples) |*t| {
        if (t.active and
            t.subject_len == s_len and bytesEqual(t.subject[0..t.subject_len], s) and
            t.predicate_len == p_len and bytesEqual(t.predicate[0..t.predicate_len], p) and
            t.object_len == o_len and bytesEqual(t.object[0..t.object_len], o))
        {
            t.active = false;
            sessions[idx].triple_count -= 1;
            return 0;
        }
    }
    return 1;
}

/// Returns total triple count.
pub export fn triplestore_count(slot: c_int) callconv(.c) u32 {
    mutex.lock();
    defer mutex.unlock();
    const idx = validSlot(slot) orelse return 0;
    return sessions[idx].triple_count;
}

/// Check if a triple exists (exact match). Returns 1 if exists, 0 if not.
pub export fn triplestore_has(
    slot: c_int,
    s_ptr: [*]const u8,
    s_len: u32,
    p_ptr: [*]const u8,
    p_len: u32,
    o_ptr: [*]const u8,
    o_len: u32,
) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 0;

    const s = s_ptr[0..s_len];
    const p = p_ptr[0..p_len];
    const o = o_ptr[0..o_len];

    for (&sessions[idx].triples) |*t| {
        if (t.active and
            t.subject_len == s_len and bytesEqual(t.subject[0..t.subject_len], s) and
            t.predicate_len == p_len and bytesEqual(t.predicate[0..t.predicate_len], p) and
            t.object_len == o_len and bytesEqual(t.object[0..t.object_len], o))
        {
            return 1;
        }
    }
    return 0;
}

/// Begin a transaction. Returns 0 on success, 1 on rejection.
/// Transitions Ready -> Transaction.
pub export fn triplestore_txn_begin(slot: c_int) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 1;
    if (sessions[idx].state != .ready) return 1;

    sessions[idx].txn_snapshot_count = sessions[idx].triple_count;
    sessions[idx].state = .transaction;
    return 0;
}

/// Commit a transaction. Returns 0 on success, 1 on rejection.
/// Transitions Transaction -> Ready.
pub export fn triplestore_txn_commit(slot: c_int) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 1;
    if (sessions[idx].state != .transaction) return 1;

    sessions[idx].state = .ready;
    return 0;
}

/// Rollback a transaction. Returns 0 on success, 1 on rejection.
/// Transitions Transaction -> Ready, reverting triple count.
pub export fn triplestore_txn_rollback(slot: c_int) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 1;
    if (sessions[idx].state != .transaction) return 1;

    // Simplified rollback: deactivate triples added during transaction
    var count: u32 = 0;
    for (&sessions[idx].triples) |*t| {
        if (t.active) {
            if (count >= sessions[idx].txn_snapshot_count) {
                t.active = false;
            } else {
                count += 1;
            }
        }
    }
    sessions[idx].triple_count = sessions[idx].txn_snapshot_count;
    sessions[idx].state = .ready;
    return 0;
}

/// Begin a bulk import. Returns 0 on success, 1 on rejection.
/// Transitions Ready -> Importing.
pub export fn triplestore_import_begin(slot: c_int, format: u8) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 1;
    if (sessions[idx].state != .ready) return 1;
    if (format > 5) return 1;

    sessions[idx].import_format = @enumFromInt(format);
    sessions[idx].state = .importing;
    return 0;
}

/// End a bulk import. Returns 0 on success, 1 on rejection.
/// Transitions Importing -> Ready.
pub export fn triplestore_import_end(slot: c_int) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 1;
    if (sessions[idx].state != .importing) return 1;

    sessions[idx].state = .ready;
    return 0;
}

/// Disconnect the store. Returns 0 on success, 1 on rejection.
pub export fn triplestore_disconnect(slot: c_int) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 1;
    if (sessions[idx].state == .ready) {
        sessions[idx].state = .closing;
        return 0;
    }
    return 1;
}

/// Complete cleanup after disconnect. Returns 0 on success, 1 on rejection.
pub export fn triplestore_cleanup(slot: c_int) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 1;
    if (sessions[idx].state != .closing) return 1;

    sessions[idx].state = .idle;
    sessions[idx].triples = [_]Triple{empty_triple} ** MAX_TRIPLES;
    sessions[idx].triple_count = 0;
    return 0;
}

/// Check if a store state transition is valid.
pub export fn triplestore_can_transition(from: u8, to: u8) callconv(.c) u8 {
    if (from == 0 and to == 1) return 1; // Idle -> Ready
    if (from == 1 and to == 2) return 1; // Ready -> Transaction
    if (from == 2 and to == 1) return 1; // Transaction -> Ready (commit/rollback)
    if (from == 1 and to == 3) return 1; // Ready -> Importing
    if (from == 3 and to == 1) return 1; // Importing -> Ready
    if (from == 1 and to == 4) return 1; // Ready -> Closing
    if (from == 4 and to == 0) return 1; // Closing -> Idle
    return 0;
}

/// Returns number of active sessions.
pub export fn triplestore_session_count() callconv(.c) u32 {
    mutex.lock();
    defer mutex.unlock();
    var count: u32 = 0;
    for (&sessions) |*s| {
        if (s.active) count += 1;
    }
    return count;
}
